File: MTGSearch/src/test/java/com/dbottillo/mtgsearchfree/interactors/SetsInteractorImplTest.java
package com.dbottillo.mtgsearchfree.interactors;

import com.dbottillo.mtgsearchfree.model.MTGSet;
import com.dbottillo.mtgsearchfree.model.database.SetDataSource;
import com.dbottillo.mtgsearchfree.util.Logger;

import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import rx.observers.TestSubscriber;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class SetsInteractorImplTest {

    @Rule
    public MockitoRule mockitoRule = MockitoJUnit.rule();

    @Mock
    Logger logger;

    private List<MTGSet> sets = Arrays.asList(new MTGSet(1, "Zendikar"), new MTGSet(2, "Ravnica"));

    @Test
    public void testLoad() throws Exception {
        SetDataSource setDataSource = mock(SetDataSource.class);
        when(setDataSource.getSets()).thenReturn(sets);
        SetsInteractor interactor = new SetsInteractorImpl(setDataSource, logger);
        TestSubscriber<List<MTGSet>> testSubscriber = new TestSubscriber<>();
        interactor.load().subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertReceivedOnNext(Collections.singletonList(sets));
        verify(setDataSource).getSets();
    }
}

File: MTGSearch/src/test/kotlin/com/dbottillo/mtgsearchfree/interactors/SetsInteractorImplTest.kt
package com.dbottillo.mtgsearchfree.interactors

import com.dbottillo.mtgsearchfree.model.MTGSet
import com.dbottillo.mtgsearchfree.database.SetDataSource
import com.dbottillo.mtgsearchfree.interactor.SchedulerProvider
import com.dbottillo.mtgsearchfree.util.Logger
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.verifyNoMoreInteractions
import com.nhaarman.mockito_kotlin.whenever

import org.junit.Rule
import org.junit.Test
import org.mockito.Mock
import org.mockito.junit.MockitoJUnit
import io.reactivex.observers.TestObserver
import io.reactivex.schedulers.Schedulers
import org.junit.Before

class SetsInteractorImplTest {

    @Rule @JvmField var mockitoRule = MockitoJUnit.rule()!!

    @Mock private lateinit var logger: Logger
    @Mock private lateinit var dataSource: SetDataSource
    @Mock private lateinit var sets: List<MTGSet>
    @Mock lateinit var schedulerProvider: SchedulerProvider

    private lateinit var underTest: SetsInteractor

    @Before
    fun setUp() {
        whenever(schedulerProvider.io()).thenReturn(Schedulers.trampoline())
        whenever(schedulerProvider.ui()).thenReturn(Schedulers.trampoline())
        underTest = SetsInteractorImpl(dataSource, schedulerProvider, logger)
    }

    @Test
    fun `load should call interactor and return observable`() {
        whenever(dataSource.sets).thenReturn(sets)

        val result = underTest.load()

        val testSubscriber = TestObserver<List<MTGSet>>()
        result.subscribe(testSubscriber)
        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(sets)
        verify(dataSource).sets
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(dataSource, schedulerProvider)
    }
}